File: services/processing-core/app/Application/Weather/DTO/WeatherSnapshotPayload.php
<?php

namespace App\Application\Weather\DTO;

/**
 * DTO koji predstavlja business payload eventa weather.snapshot.fetched.
 *
 * Ovo je tipizovana verzija payload-a koju koristi handler, kako ne bi
 * radio direktno sa sirovim nizovima.
 */
final readonly class WeatherSnapshotPayload
{
    public function __construct(
        public string $locationId,
        public string $city,
        public string $country,
        public float $latitude,
        public float $longitude,
        public float $temperatureC,
        public float $windSpeedKmh,
        public int $weatherCode,
        public string $observedAt,
        public string $source,
    ) {
    }

    /**
     * Kreira DTO iz payload niza.
     *
     * @param array<string, mixed> $payload
     */
    public static function fromArray(array $payload): self
    {
        return new self(
            locationId: (string) $payload['location_id'],
            city: (string) $payload['city'],
            country: (string) $payload['country'],
            latitude: (float) $payload['latitude'],
            longitude: (float) $payload['longitude'],
            temperatureC: (float) $payload['temperature_c'],
            windSpeedKmh: (float) $payload['wind_speed_kmh'],
            weatherCode: (int) $payload['weather_code'],
            observedAt: (string) $payload['observed_at'],
            source: (string) $payload['source'],
        );
    }

    /**
     * Vraća DTO kao niz.
     *
     * @return array<string, mixed>
     */
    public function toArray(): array
    {
        return [
            'location_id' => $this->locationId,
            'city' => $this->city,
            'country' => $this->country,
            'latitude' => $this->latitude,
            'longitude' => $this->longitude,
            'temperature_c' => $this->temperatureC,
            'wind_speed_kmh' => $this->windSpeedKmh,
            'weather_code' => $this->weatherCode,
            'observed_at' => $this->observedAt,
            'source' => $this->source,
        ];
    }
}

File: services/processing-core/src/Application/Messaging/DTO/WeatherSnapshotPayload.cs
using System.Text.Json.Serialization;
using Application.Messaging.Serialization;

namespace Application.Messaging.DTO;

/// <summary>
/// Canonical payload za weather.snapshot.fetched event.
/// </summary>
public sealed class WeatherSnapshotPayload
{
    [JsonPropertyName("location_id")]
    public Guid LocationId { get; init; }

    [JsonPropertyName("city")]
    public string City { get; init; } = string.Empty;

    [JsonPropertyName("country")]
    public string Country { get; init; } = string.Empty;

    [JsonPropertyName("latitude")]
    public decimal Latitude { get; init; }

    [JsonPropertyName("longitude")]
    public decimal Longitude { get; init; }

    [JsonPropertyName("temperature_c")]
    public decimal TemperatureC { get; init; }

    [JsonPropertyName("wind_speed_kmh")]
    public decimal WindSpeedKmh { get; init; }

    [JsonPropertyName("weather_code")]
    public int WeatherCode { get; init; }

    [JsonPropertyName("observed_at")]
    [JsonConverter(typeof(FlexibleDateTimeOffsetJsonConverter))]
    public DateTimeOffset ObservedAt { get; init; }

    [JsonPropertyName("source")]
    public string Source { get; init; } = string.Empty;
}
